File: leetcode/3/LongestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring0 = function(s) {
    let i = 0; 
    let j = 0;
    let max = 0;
    let set = new Set();
    while (i < s.length)
    {
        
        if(!set.has(s.charAt(i)))
        {
            // console.log(i,s.charAt(i),set.has(s.charAt(i)));
           set.add(s.charAt(i++));
           max = Math.max(max,set.size);     
        }
        else
        {
            // console.log(i,s.charAt(i),set.has(s.charAt(i)));
            set.delete(s.charAt(j++));
        }
    }
    return max;
};

var lengthOfLongestSubstring = function(s) {
    let i = 0;
    let j = 0;
    let max = 0;
    
    let hashset = new Set();
    
    while( i < s.length )
    {
        if(!hashset.has(s[i]))
        {
            hashset.add(s[i++]);
            max = Math.max(hashset.size, max);
        }else
        {
            hashset.delete(s[j++]);
        }
    }
    return max;
};

console.log(lengthOfLongestSubstring0("abcabcbb"));
console.log(lengthOfLongestSubstring("abcabcbb"));

File: leetcode/3/LongestSubstring.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    i = 0
    j = 0
    set = Set[];
    max_length = 0
    while i < s.length
        if !(set.include?(s[i]))
            set.add?(s[i])
            i = i + 1
            max_length = [set.size(),max_length].max
        elsif
            set.delete?(s[j])
            j = j +1
        end
    end
    max_length
end
puts(length_of_longest_substring("abcabcbb"));